text/0000-unreservations-2018.md

```diff
@@ -0,0 +1,303 @@
+made a mistake.
+
+See the rationale for potential risks with each keyword.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There's only one alternative: Not unreserving all listed / some keywords.
+
+Not unreserving a keyword would make the word unavailable for use as an
+identifier.
+
+## General policy around unreservations
+
+This RFC establishes a general rationale and policy for keyword unreservation:
+*If we are not using a keyword for anything in the language, and we are sure
+that we have no intention of using the keyword in the future, then it is
+permissible to unreserve a keyword and it is motivated.
+Additionally, if there is a desire for a keyword to be used as an identifier,
+this can in some cases outweigh very hypothetical and speculative language features.*
+
+## Rationale for `pure`
+
+This keyword used to be used for `pure fn`, that is: as an effect.
+
+[applicative]: http://hackage.haskell.org/package/base-4.11.1.0/docs/Control-Applicative.html#t:Applicative
+
+When *generic associated types* (GATs) lands, it is likely that people would
+like to use this in their [applicative functor][applicative] and monad libraries,
+which speaks in favor of unreserving `pure`. This use case explicitly mentioned by [`@ubsan`](https://github.com/ubsan/) who requested that the keyword be unreserved for this purpose.
+
+### Potential drawbacks
+
+Examples / The reasons why we might want to keep `pure` reserved are:
+
+#### 1. Effects
+
+```rust
+pure fn foo(x: Type) -> Type {
+    ...
+}
+```
+
+Here, `pure` denotes a deterministic function -- but we already have `const`
+for more or less the same, and it is unlikely that we would introduce an effect
+(or restriction thereof) that is essentially `const fn` but not entirely.
+So this use case is unlikely to happen.
+
+#### 2. Explicit *`Ok`-wrapping*
+
+```rust
+fn foo() -> Result<i32, Error> {
+    if bar() {
+        pure 0;
+    }
+    ...
+}
+```
+
+desugars into:
+
+```rust
+fn foo() -> Result<i32, Error> {
+    if bar() {
+        return Try::from_ok(0);
+    }
+    ...
+}
+```
+
+[Applicative laws]: https://en.wikibooks.org/wiki/Haskell/Applicative_functors#Applicative_functor_laws
+
+While you might think that Haskell developers would be in favor of this,
+that does not seem to be the case. Haskell developers over at
+`#haskell @ freenode` were not particularly in favor of this use as `pure`
+in this context as `pure` does not respect the [Applicative laws].
+The desugaring is also not particularly obvious when `pure` is used.
+If we did add sugar for explicit `Ok`-wrapping, we'd probably go with something
+other than `pure`.
+
+#### Summary
+
+In both 1. and 2., `pure` can be contextual.
+We also don't think that the drawbacks are significant for `pure`.
+
+## Rationale for `unsize`
+
+This would be a modifier on types, but we already have `<T: ?Sized>` and we
+could have `T: !Sized` so there seems to be no need for keeping `unsized`.
```

Comment:
nit: `unsize` or `unsized`?  It's `unsized` in the reference, but `unsize` in the header, the summary, ...

Reply:
Oops, it should be `unsized` 😆 I'll fix this when I get home later today :)